## 1. Loading Package

In [10]:
#Basic import
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import time
import pickle
from datetime import timedelta

#Machine Learning
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error,accuracy_score, classification_report, precision_score, recall_score

## 2. Data Loading

In [2]:
#Import data_mnist
data = pd.read_csv (r'/Users/joan/Desktop/Spring2022/Machine Learning 6364/HW2/data_mnist.csv')

#Import test_mnist
test= pd.read_csv(r'/Users/joan/Desktop/Spring2022/Machine Learning 6364/HW2/test_mnist.csv')

In [3]:
data.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [4]:
test.head()

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 784 columns]

## 3. Data Convert and Preprocess

##### 3.1 Convert the dataframe into array (matrix)

In [5]:
data_matrix=np.array(data)

##### 3.2 Split the data into X(features) and Y(target)

In [6]:
X=data_matrix[:,1:]
Y=data_matrix[:,0]

##### 3.3  Normalization

In [7]:
norm_X = X/ 255.0

##### 3.4 Split data into two sets: (1) Train data set (2) Test data set {Train-Test split algorithms}

In [8]:
#Both of theses two sets are from data_mnist.csv
X_train,X_test,Y_train,Y_test = train_test_split(norm_X,Y,test_size=0.25,random_state=42, stratify=Y)

## 4.Model Training

##### 4.1 Train the model when solver ="newton-cg"

In [11]:
start_time = time.time()
model_newton = LogisticRegression(solver ='newton-cg',penalty="l2",max_iter= 3000)
model_newton.fit(X_train,Y_train)

print("--- %s time elapsed ---" % str(timedelta(seconds=time.time() - start_time)))

--- 0:00:36.411517 time elapsed ---


In [12]:
y_pred = model_newton.predict(X_test)

In [13]:
class_report_newton = classification_report(Y_test, y_pred)
accuracy_newton = accuracy_score(Y_test,y_pred)*100
precision_newton = precision_score(Y_test,y_pred,average='micro')*100

In [14]:
print('Logistic Model Evaluation- "newton-cg":')
print('')
print(class_report_newton)
print('')
print("Model Score is %.4f" % model_newton.score(X_train, Y_train))
print("Accuracy is : {:.4f}%".format(accuracy_newton) )
print("Precision is : {:.4f}%".format(precision_newton) )

Logistic Model Evaluation- "newton-cg":

              precision    recall  f1-score   support

           0       0.96      0.95      0.96      1033
           1       0.93      0.98      0.95      1171
           2       0.91      0.90      0.91      1044
           3       0.90      0.89      0.89      1088
           4       0.92      0.92      0.92      1018
           5       0.86      0.86      0.86       949
           6       0.95      0.95      0.95      1034
           7       0.93      0.93      0.93      1100
           8       0.88      0.85      0.86      1016
           9       0.88      0.89      0.89      1047

    accuracy                           0.91     10500
   macro avg       0.91      0.91      0.91     10500
weighted avg       0.91      0.91      0.91     10500


Model Score is 0.9469
Accuracy is : 91.3333%
Precision is : 91.3333%


##### 4.2 Store the model_newton

In [15]:
pickle.dump(model_newton, open("Model_newton", "wb"))

##### 4.3 Train the model when solver = "sag"

In [16]:
start_time = time.time()
model_sag = LogisticRegression(solver ='sag',penalty="l2",max_iter= 3000)
model_sag.fit(X_train,Y_train)

print("--- %s time elapsed ---" % str(timedelta(seconds=time.time() - start_time)))

--- 0:03:51.834441 time elapsed ---


In [17]:
y_pred = model_sag.predict(X_test)

In [18]:
class_report_sag = classification_report(Y_test, y_pred)
accuracy_sag = accuracy_score(Y_test,y_pred)*100
precision_sag = precision_score(Y_test,y_pred,average='micro')*100

In [19]:
print('Logistic Model Evaluation- "sag":')
print('')
print(class_report_sag)
print('')
print("Model Score is %.4f" % model_sag.score(X_train, Y_train))
print("Accuracy is : {:.4f}%".format(accuracy_sag) )
print("Precision is : {:.4f}%".format(precision_sag) )

Logistic Model Evaluation- "sag":

              precision    recall  f1-score   support

           0       0.96      0.95      0.96      1033
           1       0.93      0.98      0.95      1171
           2       0.91      0.90      0.91      1044
           3       0.90      0.89      0.89      1088
           4       0.92      0.92      0.92      1018
           5       0.86      0.86      0.86       949
           6       0.95      0.95      0.95      1034
           7       0.93      0.93      0.93      1100
           8       0.88      0.85      0.86      1016
           9       0.88      0.89      0.89      1047

    accuracy                           0.91     10500
   macro avg       0.91      0.91      0.91     10500
weighted avg       0.91      0.91      0.91     10500


Model Score is 0.9469
Accuracy is : 91.3333%
Precision is : 91.3333%


##### 4.4 Store the model_sag

In [20]:
pickle.dump(model_sag, open("Model_sag", "wb"))

##### 4.5 Train the model when solver ="saga"¶

In [23]:
start_time = time.time()
model_saga = LogisticRegression(solver ='saga',penalty='l2',max_iter= 3000)
model_saga.fit(X_train,Y_train)

print("--- %s time elapsed ---" % str(timedelta(seconds=time.time() - start_time)))

--- 0:07:26.021090 time elapsed ---


In [24]:
y_pred = model_saga.predict(X_test)

In [27]:
class_report_saga = classification_report(Y_test, y_pred)
accuracy_saga = accuracy_score(Y_test,y_pred)*100
precision_saga = precision_score(Y_test,y_pred,average='micro')*100

In [28]:
print('Logistic Model Evaluation- "saga":')
print('')
print(class_report_saga)
print('')
print("Model Score is %.4f" % model_sag.score(X_train, Y_train))
print("Accuracy is : {:.4f}%".format(accuracy_saga) )
print("Precision is : {:.4f}%".format(precision_saga) )

Logistic Model Evaluation- "saga":

              precision    recall  f1-score   support

           0       0.96      0.95      0.96      1033
           1       0.93      0.98      0.95      1171
           2       0.92      0.90      0.91      1044
           3       0.90      0.89      0.89      1088
           4       0.92      0.92      0.92      1018
           5       0.86      0.86      0.86       949
           6       0.95      0.96      0.95      1034
           7       0.93      0.93      0.93      1100
           8       0.88      0.85      0.86      1016
           9       0.88      0.89      0.89      1047

    accuracy                           0.91     10500
   macro avg       0.91      0.91      0.91     10500
weighted avg       0.91      0.91      0.91     10500


Model Score is 0.9469
Accuracy is : 91.3524%
Precision is : 91.3524%


##### 4.6 Store the model_saga

In [29]:
pickle.dump(model_saga, open("Model_saga", "wb"))

##### 4.7 Train the model when solver ="lbfgs"¶

In [30]:
start_time = time.time()
model_lbfgs = LogisticRegression(solver ='lbfgs',penalty='l2',max_iter= 3000)
model_lbfgs.fit(X_train,Y_train)

print("--- %s time elapsed ---" % str(timedelta(seconds=time.time() - start_time)))

--- 0:00:36.704475 time elapsed ---


In [31]:
y_pred = model_lbfgs.predict(X_test)

In [32]:
class_report_lbfgs = classification_report(Y_test, y_pred)
accuracy_lbfgs = accuracy_score(Y_test,y_pred)*100
precision_lbfgs = precision_score(Y_test,y_pred,average='micro')*100

In [34]:
print('Logistic Model Evaluation- "lbfgs":')
print('')
print(class_report_lbfgs)
print('')
print("Model Score is %.4f" % model_sag.score(X_train, Y_train))
print("Accuracy is : {:.4f}%".format(accuracy_lbfgs) )
print("Precision is : {:.4f}%".format(precision_lbfgs) )

Logistic Model Evaluation- "lbfgs":

              precision    recall  f1-score   support

           0       0.96      0.95      0.96      1033
           1       0.93      0.98      0.95      1171
           2       0.91      0.90      0.91      1044
           3       0.90      0.89      0.89      1088
           4       0.92      0.92      0.92      1018
           5       0.86      0.86      0.86       949
           6       0.95      0.95      0.95      1034
           7       0.93      0.93      0.93      1100
           8       0.88      0.85      0.87      1016
           9       0.88      0.89      0.89      1047

    accuracy                           0.91     10500
   macro avg       0.91      0.91      0.91     10500
weighted avg       0.91      0.91      0.91     10500


Model Score is 0.9469
Accuracy is : 91.3429%
Precision is : 91.3429%


##### 4.8 Store the model_lbfgs

In [35]:
pickle.dump(model_saga, open("Model_lbfgs", "wb"))

### How to evaluate the performance...

#### To evaluate the model, we usually use **the accuracy** to evaluate the model's performance. 

In Logistic Regression Model, there are different parameters that will affect the model's accuracy and it's performance. One of the parameters is the "solver", there are five solvers in this model that we can choose to use, including ‘newton-cg’, ‘lbfgs’, ‘liblinear’, ‘sag’, ‘saga’ and the default setting is 'lbfgs'.
In this exercise, since the label set consists of ten possible digits, therefore, we need to building a multi-class logisitc regression classifier.

For multi-class problems in logistic regression model, only **‘newton-cg’, ‘sag’, ‘saga’ and ‘lbfgs’** handle multinomial loss. And that's the reason why in the above model training, the parameters that we choose for solver are ‘newton-cg’,‘sag’, ‘saga’ and ‘lbfgs’.

After training these four different solver, we can tell from the results that when solver is "saga", **it will have the higher accuracy than the others.** Even it's training time is the longest, it has the best accuracy. 

#### Therefore, in the following prediction, we will use the model that the parameter of solver is saga.

## 5. Predict 

##### 5.1 Load the Model_saga

In [36]:
#Load the Model_saga
LogisticReg_Model=pickle.load(open("Model_saga", 'rb'))

##### 5.2 Convert the test.csv dataframe into array

In [37]:
test_matrix=np.array(test)

##### 5.3 Predict Model_saga

In [38]:
test_pred_LogisticReg=LogisticReg_Model.predict(test_matrix)

In [39]:
#Convert the array into dataframe
pred_DF= pd.DataFrame({'ImageId':np.arange(len(test)),'Label':test_pred_LogisticReg})

In [41]:
test['Label'] = test_pred_LogisticReg
# print (test.head(5))

## 6. Submission

In [42]:
print("The Final Result for the test file prediction:")
print(pred_DF)

The Final Result for the test file prediction:
      ImageId  Label
0           0      7
1           1      2
2           2      1
3           3      0
4           4      4
...       ...    ...
9995     9995      2
9996     9996      3
9997     9997      4
9998     9998      5
9999     9999      6

[10000 rows x 2 columns]


In [43]:
pred_DF.to_csv('hw2_submission.csv', index=False)

## 7. Compare to kNN MNIST classifier model 

**The reason why we need to split the data and preprocess the data again is because the way we preprocess in the Logistic regression model is different than the KNN model.**
1. In Logisitc Regression Model, the preprocess is using the "Normalization" which we divide each element in a sigle data grid by 255.
2. In KNN Model, the preprocess is using the "StandardScaler".

**Therefore, to get the precise accuracy that we train in the best KNN model, we need to use the same way of preprocess which is "StandardScaler" to predict the data and calculate the accuracy.**

##### 7.1 Load the KNN Best Model

In [44]:
KNN_Model=pickle.load(open("Best_model", 'rb'))

##### 7.2 Split data into two sets: (1) Train data set (2) Test data set

In [45]:
X_train_KNN,X_test_KNN,Y_train_KNN,Y_test_KNN = train_test_split(X,Y,test_size=0.25,random_state=42, stratify=Y)

##### 7.3 Data Preprocessing- StandardScaler

In [46]:
sc_X = StandardScaler()

sc_X.fit(X_train_KNN)
X_train_std=sc_X.transform(X_train_KNN)
X_test_std=sc_X.transform(X_test_KNN)

##### 7.4 Predict using KNN Model

In [47]:
y_pred_KNN = KNN_Model.predict(X_test_std)

In [48]:
class_report = classification_report(Y_test_KNN, y_pred_KNN)
accuracy = accuracy_score(Y_test_KNN,y_pred_KNN)*100
precision = precision_score(Y_test_KNN,y_pred_KNN,average='micro')*100

In [49]:
print('KNN Model Evaluation:')
print('')
print(class_report)
print('')
print("Model Score is %.4f" % KNN_Model.score(X_train_KNN, Y_train_KNN))
print("Accuracy is : {:.4f}%".format(accuracy) )
print("Precision is : {:.4f}%".format(precision) )

KNN Model Evaluation:

              precision    recall  f1-score   support

           0       0.96      0.98      0.97      1033
           1       0.95      0.99      0.97      1171
           2       0.95      0.94      0.94      1044
           3       0.91      0.93      0.92      1088
           4       0.95      0.92      0.93      1018
           5       0.92      0.91      0.92       949
           6       0.95      0.98      0.96      1034
           7       0.93      0.93      0.93      1100
           8       0.97      0.87      0.92      1016
           9       0.90      0.93      0.92      1047

    accuracy                           0.94     10500
   macro avg       0.94      0.94      0.94     10500
weighted avg       0.94      0.94      0.94     10500


Model Score is 0.8614
Accuracy is : 93.9333%
Precision is : 93.9333%


##### 7.5 Predicting using Logistic Model -saga

In [50]:
y_pred = model_saga.predict(X_test)

In [51]:
class_report_saga = classification_report(Y_test, y_pred)
accuracy_saga = accuracy_score(Y_test,y_pred)*100
precision_saga = precision_score(Y_test,y_pred,average='micro')*100

In [53]:
print('Logistic Model Evaluation- "saga":')
print('')
print(class_report_saga)
print('')
print("Model Score is %.4f" % model_sag.score(X_train, Y_train))
print("Accuracy is : {:.4f}%".format(accuracy_saga) )
print("Precision is : {:.4f}%".format(precision_saga) )

Logistic Model Evaluation- "saga":

              precision    recall  f1-score   support

           0       0.96      0.95      0.96      1033
           1       0.93      0.98      0.95      1171
           2       0.92      0.90      0.91      1044
           3       0.90      0.89      0.89      1088
           4       0.92      0.92      0.92      1018
           5       0.86      0.86      0.86       949
           6       0.95      0.96      0.95      1034
           7       0.93      0.93      0.93      1100
           8       0.88      0.85      0.86      1016
           9       0.88      0.89      0.89      1047

    accuracy                           0.91     10500
   macro avg       0.91      0.91      0.91     10500
weighted avg       0.91      0.91      0.91     10500


Model Score is 0.9469
Accuracy is : 91.3524%
Precision is : 91.3524%


### Conclude the comparison of KNN model and Logistic Regression model...

#### KNN model has higher accuracy than Logistic Regressoon model
1. The Accuracy of the KNN model is "93.9333 %" 
2. The Accuracy of the Logistic Regression model is "91.3524 %"

#### Therefore, from the above result we can tell that KNN model has higher performance than Logistic Regressoon model.